cli/src/index.ts

```diff
@@ -15,63 +23,249 @@ export type RootCommand = typeof rootCommand;
       alias: "cs",
       type: "string",
       describe: "The Jira consumer secret in base64 format",
+      default: config?.jiraConsumerSecret,
     })
-    .group(["jiraConsumerKey", "jiraConsumerSecret"], authOptionGroup)
-    .demandOption(["jiraHost", "jiraConsumerSecret"]);
+    .group([jiraHostOptionKey, jiraProtocolOptionKey], "Common Required:");
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export const withAuthenticationOptions = <C extends RootCommand>(command: C) =>
+  withCommonOptions(command)
+    .group([jiraConsumerKeyOptionKey, jiraConsumerSecretOptionKey], "Auth:")
+    .demandOption([jiraHostOptionKey, jiraConsumerSecretOptionKey]);
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
 export const withAuthOptions = <C extends RootCommand>(command: C) =>
   withCommonOptions(command)
-    .option("accessToken", {
+    .option(jiraAccessTokenOptionKey, {
       alias: "t",
       type: "string",
       describe: "access token",
+      default: config?.accessToken,
     })
-    .option("accessSecret", {
+    .option(jiraAccessSecretOptionKey, {
       alias: "s",
       type: "string",
       describe: "access secret",
+      default: config?.accessSecret,
+    })
+    .option(jiraUsernameOptionKey, {
+      alias: "u",
+      type: "string",
+      describe: "your jira username",
+      default: config?.username,
+    })
+    .option(jiraPasswordOptionKey, {
+      alias: "z",
+      type: "string",
+      describe: "your jira password",
+      default: config?.password,
+    })
+    .option(jiraPersonalAccessTokenOptionKey, {
+      alias: "k",
+      type: "string",
+      describe: "your jira personal access token",
+      default: config?.personalAccessToken,
+    })
+    .group(
+      [
+        jiraConsumerKeyOptionKey,
+        jiraConsumerSecretOptionKey,
+        jiraAccessTokenOptionKey,
+        jiraAccessSecretOptionKey,
+      ],
+      "Connection - OAuth:"
+    )
+    .group(
+      [jiraUsernameOptionKey, jiraPasswordOptionKey],
+      "Connection - User Credentials:"
+    )
+    .group(
+      [jiraPersonalAccessTokenOptionKey],
+      "Connection - Personal Access Token:"
+    )
+    .option("jira", {
+      describe: "Provide at least one full set of connection parameters",
     })
-    .group(["accessToken", "accessSecret"], authOptionGroup)
-    .demandOption(["accessToken", "accessSecret"]);
+    .coerce(
+      "jira",
+      (jira: {
+        readonly protocol?: "http" | "https";
+        readonly host?: string;
+        readonly consumerKey?: string;
+        readonly consumerSecret?: string;
+        readonly accessToken?: string;
+        readonly accessSecret?: string;
+        readonly personalAccessToken?: string;
+        readonly username?: string;
+        readonly password?: string;
+      }): JiraClient => {
+        // eslint-disable-next-line functional/no-conditional-statement
+        if (jira.protocol === undefined) {
```

Comment:
I tried turning this into a ternary, but it makes it much more complex to read and eslint fails on the function complexity.

Reply:
Can `yargs` guarantee that this required fields are defined and get rid of this entirely?

Reply:
Unfortunately not. They need to be included in the jira object to use them in coerce (so --jira.host) but as soon as we do that, making them required breaks coerce. I tried to find a way around this but was unsuccessful. I'll have another look as I change this function to collect the errors.